### PR TITLE
Settings page: Improve Site ID handling

### LINKED ISSUE #763: Settings page: Improve site ID validation/sanitization
## Is your feature request related to a problem?
@ryanwelcher reported that it is possible to insert `https://` in front of the site ID, as well as trailing `/` in the plugin's settings. This passes validation but it's not a valid site ID format. And although we do have the help text below, it can happen that users don't notice.

## Describe the solution you'd like
- Add `yourdomain.com` as a placeholder to the site ID textbox
- Possibly enhance the help text to make it more explicit
- Look out for, or automatically remove `http://`, `https://`, `www.` and possible trailing `/`. I'm currently checking with Parse.ly to see if any of the prefixes could be valid for a site ID.

### PR DESCRIPTION
**Pull request by acicovic**
## Description
Our settings page was allowing to save certain invalid Site IDs (for example strings containing `http(s)://` or trailing slashes) which in turn resulted into problems. Also, the help text's wording had some room for improvement.

This PR improves the help text as to be as precise as possible, but also imposes a bigger restriction on what is considered as a valid Site ID. It does accept a `www` prefix that is used in certain keys, and can accept strings with up to 3 `.` characters.

For example:
- `test.com`: valid
- `test.com/`: **invalid**
- `http://test.com`: **invalid**
- `https://test.com`: **invalid**
- `www.test.com`: valid
- `subdomain.test.com`: valid
- `www.subdomain.test.com`: valid
- `subdomain.subdomain.test.com`: valid
- `www.subdomain.subdomain.test.com`: **invalid**

## Motivation and Context
- Be clearer on how we expect Site IDs to be.
- Avoid problems by further limiting the possibility of accepting invalid site IDs.
- Fix #763.

## How Has This Been Tested?
- Manual testing.
- Unit tests created.

### CODE REVIEW

File: src/UI/class-settings-page.php [hunk 1]
```diff
@@ -909,9 +910,9 @@ public function validate_options( array $input ): array {
 				__( 'Please specify the Site ID', 'wp-parsely' )
 			);
 		} else {
-			$input['apikey'] = strtolower( $input['apikey'] );
-			$input['apikey'] = sanitize_text_field( $input['apikey'] );
-			if ( strpos( $input['apikey'], '.' ) === false || strpos( $input['apikey'], ' ' ) !== false ) {
+			$api_key_format  = '/^((\w+)\.)?(([\w-]+)?)(\.[\w-]+){1,2}$/';
+			$input['apikey'] = strtolower( sanitize_text_field( $input['apikey'] ) );
+			if ( 1 !== preg_match( $api_key_format, $input['apikey'] ) ) {
```
Comment (pauarge) [inline src/UI/class-settings-page.php]: Can we extract the check of a correct API key to a separate function?

**Review by pauarge — CHANGES_REQUESTED** [review]

**Review by GaryJones — DISMISSED** [review]
This is going to need unit tests for the validation of the string.

File: src/UI/class-settings-page.php [hunk 2]
```diff
@@ -310,14 +310,15 @@ private function initialize_basic_section(): void {
 		);
 
 		// Get the API Key.
-		$h          = __( 'Your Site ID is your own site domain (e.g. `mydomain.com`).', 'wp-parsely' );
+		$h          = __( 'Your Site ID is your own site domain without <code>http(s)://</code> prefixes or trailing <code>/</code> (e.g. <code>mydomain.com</code>).', 'wp-parsely' );
```
Comment (GaryJones) [inline src/UI/class-settings-page.php]:
```suggestion
		$h          = __( 'Your Site ID is typically your own site domain without <code>http(s)://</code> prefixes or trailing <code>/</code> (e.g. <code>mydomain.com</code>).', 'wp-parsely' );
```

The string could actually be anything, so while the default is to use the domain, we can allow a little wiggle room by adding a "typically".

**Comment by acicovic** [comment]
I know that [this](https://github.com/Parsely/wp-parsely/blob/0e3b1276425589c2e9249abf9d617c4f0a4f9b8d/tests/Unit/SettingsPage/ValidationsTest.php#L83-L97) is duplicated code. I have created an issue for this in #784.

**Review by pauarge — APPROVED** [review]

File: tests/Unit/SettingsPage/ValidationsTest.php [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+	private static $settings_page;
+
+	/**
+	 * Runs the setup process before each test.
+	 */
+	public function set_up(): void {
+		parent::set_up();
+
+		self::$parsely       = new Parsely();
+		self::$settings_page = new Settings_Page( self::$parsely );
```
Comment (GaryJones) [inline tests/Unit/SettingsPage/ValidationsTest.php:42]: Instantiating a class in the `UI` namespace, to test logic around a string, seems like too much coupling going on. 

Seems like perhaps the validation of the string should be handled by a validation-specific class, outside of the `UI` namespace, which the class in the `UI` namespace can then have injected, and defer to.

Comment (acicovic) [reply tests/Unit/SettingsPage/ValidationsTest.php:42]: @GaryJones, thank you. You definitely have a point here. Related to #323 and #485.

Comment (acicovic) [reply tests/Unit/SettingsPage/ValidationsTest.php:42]: I've put a mention of this comment in the above issues. Marking this conversation as resolved, as I think that changing this is part of a bigger chunk of work outside of this PR's scope.

**Pull request merged**